src/MDMTransaction.m

```diff
@@ -30,17 +32,65 @@ - (instancetype)init {
 }
 
 - (void)addPlan:(NSObject<MDMPlan> *)plan toTarget:(id)target {
-  MDMTransactionLog *log = [MDMTransactionLog new];
-  log.plans = @[ [plan copy] ];
-  log.target = target;
+  [self commonAddPlan:plan toTarget:target withName:nil];
+}
+
+- (void)addPlan:(NSObject<MDMPlan> *)plan toTarget:(id)target withName:(NSString *)name {
+  [self removePlanNamed:name fromTarget:target];
+  [self commonAddPlan:plan toTarget:target withName:name];
+}
+
+- (void)commonAddPlan:(NSObject<MDMPlan> *)plan toTarget:(id)target withName:(NSString *)name {
+  MDMTransactionLogType transactionLogType = name == nil ? MDMTransactionLogTypeUncategorized : MDMTransactionLogTypeAddNamedPlan;
+  MDMTransactionLog *log = [[MDMTransactionLog alloc] initWithPlan:[plan copy] target:target name:name transactionLogType:transactionLogType];
   [_logs addObject:log];
 }
 
+- (void)removePlanNamed:(nonnull NSString *)name fromTarget:(nonnull id)target {
+  MDMTransactionLog *log = [[MDMTransactionLog alloc] initWithPlan:nil target:target name:name transactionLogType:MDMTransactionLogTypeRemoveNamedPlan];
+  [_logs removeObject:log];
+}
+
 - (NSArray<MDMTransactionLog *> *)logs {
   return _logs;
 }
 
 @end
 
 @implementation MDMTransactionLog
+
+- (instancetype)initWithPlan:(NSObject<MDMPlan> *)plan target:(id)target name:(NSString *)name transactionLogType:(MDMTransactionLogType)transactionLogType {
+  self = [super init];
+  if (self) {
+    if (plan != nil) {
+      _plans = @[ plan ];
+    }
+    _target = target;
+    _name = [name copy];
+    _transactionLogType = transactionLogType;
+  }
+  return self;
+}
+
+- (BOOL)isEqual:(id)other {
```

Comment:
We likely don't need this method implementation if we ensure that all of the operations on the transaction are stored in the transaction log and then committed to the performer group.